src/main/java/com/google/devtools/build/lib/analysis/RunfilesSupport.java

```diff
@@ -495,4 +521,56 @@ public static Path inputManifestPath(Path runfilesDir) {
   public static Path outputManifestPath(Path runfilesDir) {
     return runfilesDir.getRelative(OUTPUT_MANIFEST_BASENAME);
   }
+
+  @Nullable
+  private static Artifact createRepoMappingManifestAction(
+      RuleContext ruleContext, Artifact owningExecutable) {
+    NestedSet<Package> transitivePackages =
+        ruleContext.getTransitivePackagesForRunfileRepoMappingManifest();
+    if (transitivePackages == null) {
+      // For environments where transitive packages are not tracked, we don't have external repos,
+      // so don't build the repo mapping manifest in such cases.
+      return null;
+    }
+    Set<NestedSet.Node> visitedNestedSets = new HashSet<>();
+    Map<RepositoryName, RepositoryMapping> reposAndMappings = new HashMap<>();
+    collectReposAndMappings(transitivePackages, visitedNestedSets, reposAndMappings);
+
+    PathFragment executablePath =
+        (owningExecutable != null)
+            ? owningExecutable.getOutputDirRelativePath(
+                ruleContext.getConfiguration().isSiblingRepositoryLayout())
+            : ruleContext.getPackageDirectory().getRelative(ruleContext.getLabel().getName());
+    Artifact repoMappingManifest =
+        ruleContext.getDerivedArtifact(
+            executablePath
+                .replaceName(executablePath.getBaseName() + RUNFILES_DIR_EXT)
+                .getRelative(REPO_MAPPING_MANIFEST_BASENAME),
+            ruleContext.getBinDirectory());
+    ruleContext
+        .getAnalysisEnvironment()
+        .registerAction(
+            new RepoMappingManifestAction(
+                ruleContext.getActionOwner(),
+                repoMappingManifest,
+                reposAndMappings,
+                ruleContext.getWorkspaceName()));
+    return repoMappingManifest;
+  }
+
+  private static void collectReposAndMappings(
```

Comment:
This is delightfully optimal code, but why not just call `toList()`? I understand that it is less optimal, but it's also much simpler and if you do it while doing I/O anyway, it probably doesn't matter and looking at the call sites of `getLeaves()`, it looks like it's intended for places more deeply in the bowels of Bazel. In particular, the comment "may wait for a future" in `getNonLeaves()` scared me.

Reply:
done -- switched to toList()